src/twisted/mail/maildir.py

```diff
@@ -261,12 +264,13 @@ def startMessage(self, user):
             name, domain = user.split('@', 1)
         else:
             name, domain = user.dest.local, user.dest.domain
-        dir = self.userDirectory(name)
+        dirPath = FilePath(self.userDirectory(name))
         fname = _generateMaildirName()
-        filename = os.path.join(dir, 'tmp', fname)
-        fp = open(filename, 'w')
-        return MaildirMessage('%s@%s' % (name, domain), fp, filename,
-                              os.path.join(dir, 'new', fname))
+        filename = dirPath.child('tmp').child(fname).asBytesMode().path
+        fp = open(filename, 'wb')
+        return MaildirMessage(
+            u'{}@{}'.format(name, domain).encode("ascii"), fp, filename,
```

Comment:
`name`/`domain` may be either `str` or `bytes` here; this will produce garbled results if they are `bytes`